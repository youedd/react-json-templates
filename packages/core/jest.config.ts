import type { Config } from 'jest'
import baseConfig from '../../jest.config.base'

const config: Config = {
  ...baseConfig,
  displayName: 'core'
}

export default config
