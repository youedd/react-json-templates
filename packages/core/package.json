{
  "name": "@react-json-templates/core",
  "version": "0.2.0-RC.9",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "license": "MIT",
  "scripts": {
    "lint": "eslint .",
    "test": "jest",
    "build": "rm -rf ./dist && tsc"
  },
  "peerDependencies": {
    "react": "*"
  }
}
