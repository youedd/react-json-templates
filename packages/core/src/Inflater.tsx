import { createContext, useContext, useMemo } from 'react'
import { type FIXME, type RJT_CONSTANT, type RJT_ELEMENT, type RJT_GENERIC_OPERATION, type RJT_OPERAND_TYPE } from './types'

export interface RJT_CONFIG {
  components: Record<string, FIXME>
  actions: Record<string, (...args: any[]) => void>
  constants: Record<string, unknown>
}

const context = createContext<RJT_CONFIG | null>(null)
export const RJTProvider = ({ config, children }: { config: RJT_CONFIG, children: any }): JSX.Element => {
  return (
    <context.Provider value={config}>
      {children}
    </context.Provider>
  )
}

export const useRJTConfig = (): RJT_CONFIG => {
  const config = useContext(context)
  if (config === null) {
    throw new Error('RJT is not configured, make sure to use this hook inside a RJTProvider !')
  }

  return config
}

export const Inflater = ({ sdui }: { sdui: RJT_ELEMENT }): FIXME => {
  const config = useRJTConfig()

  return useMemo(() => mapSDUI(sdui, config), [config, sdui])
}

const mapSDUI = (
  sdui: RJT_ELEMENT | RJT_CONSTANT | RJT_GENERIC_OPERATION | RJT_OPERAND_TYPE,
  config: RJT_CONFIG,
  key?: number
): FIXME => {
  if (Array.isArray(sdui)) {
    return sdui.map((item, index) => mapSDUI(item, config, index))
  }

  if (typeof sdui === 'object' && sdui !== null) {
    switch (sdui.type) {
      case '__RJT_FRAGMENT__': {
        return (
          <>
            {sdui.children.map((item, index) => mapSDUI(item, config, index))}
          </>
        )
      }
      case '__RJT_COMPONENT__': {
        const Comp = config.components[sdui.name]
        if (Comp == null) {
          throw new Error(`SDUI Component ${sdui.name} not found in config !`)
        }
        const props = mapSDUI(sdui.props as FIXME, config)
        return <Comp key={key} {...props} />
      }
      case '__RJT_ACTION__': {
        const action = config.actions[sdui.name]
        if (action == null) {
          throw new Error(`SDUI Action ${sdui.name} not found in config !`)
        }
        return () => { action(...sdui.params) }
      }
      case '__RJT_CONSTANT__': {
        const value = config.constants[sdui.name]
        return value
      }
      case '__RJT_OPERATION__': {
        const operation = sdui.operation
        switch (operation) {
          case '+': {
            const a = mapSDUI(sdui.operands[0], config) as number
            const b = mapSDUI(sdui.operands[1], config) as number
            return a + b
          }
          case '*': {
            return mapSDUI(sdui.operands[0], config) * mapSDUI(sdui.operands[1], config)
          }
          case '-': {
            return mapSDUI(sdui.operands[0], config) - mapSDUI(sdui.operands[1], config)
          }
          case '/': {
            return mapSDUI(sdui.operands[0], config) / mapSDUI(sdui.operands[1], config)
          }
          case 'pow': {
            return Math.pow(mapSDUI(sdui.operands[0], config), mapSDUI(sdui.operands[1], config))
          }
          case 'sqrt': {
            return Math.sqrt(mapSDUI(sdui.operands[0], config))
          }
          default: {
            throw new Error(`Unsupported operation type ${operation as string}`)
          }
        }
      }
      default: {
        const obj: Record<string, unknown> = {}

        for (const [key, value] of Object.entries(sdui)) {
          obj[key] = mapSDUI(value as RJT_ELEMENT, config)
        }

        return obj
      }
    }
  }

  return sdui
}
