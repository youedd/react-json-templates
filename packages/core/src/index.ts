export * from './types'
export * from './builders'
