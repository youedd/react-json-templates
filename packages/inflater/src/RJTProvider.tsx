import { type FIXME } from '@react-json-templates/core'
import { createContext, useContext } from 'react'

export interface RJT_CONFIG {
  components: Record<string, FIXME>
  actions: Record<string, (...args: any[]) => void>
  constants: Record<string, unknown>
}

export const RJTContext = createContext<RJT_CONFIG | null>(null)

export const RJTProvider = ({ config, children }: { config: RJT_CONFIG, children: any }): JSX.Element => {
  return (
    <RJTContext.Provider value={config}>
      {children}
    </RJTContext.Provider>
  )
}

export const useRJTConfig = (): RJT_CONFIG => {
  const config = useContext(RJTContext)
  if (config === null) {
    throw new Error('RJT is not configured, make sure to use this hook inside a RJTProvider !')
  }

  return config
}
