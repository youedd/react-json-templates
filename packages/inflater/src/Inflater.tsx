import { Fragment, useMemo } from 'react'
import type { FIXME, RJT_CONSTANT, RJT_ELEMENT, RJT_GENERIC_OPERATION, RJT_OPERAND_TYPE } from '@react-json-templates/core'
import { type RJT_CONFIG, useRJTConfig } from './RJTProvider'

export const Inflater = ({ data }: { data: RJT_ELEMENT }): FIXME => {
  const config = useRJTConfig()

  return useMemo(() => mapData(data, config), [config, data])
}

export const mapData = (
  data: RJT_ELEMENT | RJT_CONSTANT | RJT_GENERIC_OPERATION | RJT_OPERAND_TYPE,
  config: RJT_CONFIG,
  key?: number
): FIXME => {
  if (Array.isArray(data)) {
    return data.map((item, index) => mapData(item, config, index))
  }

  if (typeof data === 'object' && data !== null) {
    switch (data.type) {
      case '__RJT_FRAGMENT__': {
        return (
          <Fragment key={key}>
            {mapData(data.children, config, key)}
          </Fragment>
        )
      }
      case '__RJT_COMPONENT__': {
        const Comp = config.components[data.name]
        if (Comp == null) {
          throw new Error(`Component ${data.name} not found in config !`)
        }
        const props = mapData(data.props as FIXME, config)
        return <Comp key={key} {...props} />
      }
      case '__RJT_ACTION__': {
        const action = config.actions[data.name]
        if (action == null) {
          throw new Error(`Action ${data.name} not found in config !`)
        }
        return () => { action(...data.params) }
      }
      case '__RJT_CONSTANT__': {
        const value = config.constants[data.name]
        return value
      }
      case '__RJT_OPERATION__': {
        const operation = data.operation
        switch (operation) {
          case '+': {
            const a = mapData(data.operands[0], config) as number
            const b = mapData(data.operands[1], config) as number
            return a + b
          }
          case '*': {
            return mapData(data.operands[0], config) * mapData(data.operands[1], config)
          }
          case '-': {
            return mapData(data.operands[0], config) - mapData(data.operands[1], config)
          }
          case '/': {
            return mapData(data.operands[0], config) / mapData(data.operands[1], config)
          }
          case 'pow': {
            return Math.pow(mapData(data.operands[0], config), mapData(data.operands[1], config))
          }
          case 'sqrt': {
            return Math.sqrt(mapData(data.operands[0], config))
          }
          default: {
            throw new Error(`Unsupported operation${operation as string}`)
          }
        }
      }
      default: {
        const obj: Record<string, unknown> = {}

        for (const [key, value] of Object.entries(data)) {
          obj[key] = mapData(value as RJT_ELEMENT, config)
        }

        return obj
      }
    }
  }

  return data
}
