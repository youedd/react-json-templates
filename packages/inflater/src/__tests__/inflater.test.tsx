/**
 * @jest-environment ../../../../node_modules/jest-environment-jsdom
 */

import { render } from '@testing-library/react'
import { Serializable } from '@react-json-templates/core'
import { RJTProvider, type RJT_CONFIG } from '../RJTProvider'
import { Inflater, mapData } from '../Inflater'

const config: RJT_CONFIG = {
  components: {
    Simple: Serializable('Simple', (props: any) => <button {...props} > Simple </button>),
    Complex: Serializable('Complex', (props: any) => <div {...props} />)
  },
  actions: {
    handlePress: jest.fn(),
    renderAction: jest.fn()
  },
  constants: {
    windowWidth: 900,
    windowHeight: 1600
  }
}

describe('Inflater', () => {
  afterEach(() => {
    jest.resetAllMocks()
  })

  it('inflates simple Component', () => {
    const { asFragment } = render(
      <RJTProvider config={config} >
        <Inflater
          data={{
            type: '__RJT_COMPONENT__',
            name: 'Simple',
            props: {
              testid: 'test'
            }
          }}
        />
      </RJTProvider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('inflates complex Component', () => {
    const { asFragment } = render(
      <RJTProvider config={config} >
        <Inflater
          data={{
            type: '__RJT_COMPONENT__',
            name: 'Complex',
            props: {
              testid: 'test',
              children: [
                {
                  type: '__RJT_COMPONENT__',
                  name: 'Simple',
                  props: {}
                },
                {
                  type: '__RJT_COMPONENT__',
                  name: 'Simple',
                  props: {}
                }
              ]
            }
          }}
        />
      </RJTProvider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('inflates fragment', () => {
    const { asFragment } = render(
      <RJTProvider config={config} >
        <Inflater
          data={
            {
              type: '__RJT_COMPONENT__',
              name: 'Complex',
              props: {
                children: [
                  {
                    type: '__RJT_COMPONENT__',
                    name: 'Simple',
                    props: {}
                  },
                  {
                    type: '__RJT_FRAGMENT__',
                    children: [
                      {
                        type: '__RJT_COMPONENT__',
                        name: 'Simple',
                        props: {}
                      },
                      {
                        type: '__RJT_COMPONENT__',
                        name: 'Simple',
                        props: {}
                      }
                    ]
                  }
                ]
              }
            }
          }
        />
      </RJTProvider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('inflates props Action', () => {
    const element = mapData(
      {
        type: '__RJT_COMPONENT__',
        name: 'Simple',
        props: {
          onClick: {
            type: '__RJT_ACTION__',
            name: 'handlePress',
            params: ['arg1']
          }
        }
      },
      config
    )
    element.props.onClick()
    expect(config.actions.handlePress).toBeCalledWith('arg1')
    const { asFragment } = render(element)
    expect(asFragment()).toMatchSnapshot()
  })

  it('inflates constants', () => {
    const element = mapData(
      {
        type: '__RJT_COMPONENT__',
        name: 'Complex',
        props: {
          height: {
            type: '__RJT_CONSTANT__',
            name: 'windowHeight'
          },
          children: {
            type: '__RJT_CONSTANT__',
            name: 'windowWidth'
          }
        }
      },
      config
    )
    const { asFragment } = render(element)
    expect(asFragment()).toMatchSnapshot()
  })

  it('inflates front side calculation', () => {
    const element = mapData(
      {
        type: '__RJT_COMPONENT__',
        name: 'Simple',
        props: {
          height: {
            type: '__RJT_OPERATION__',
            operation: '+',
            operands: [
              {
                type: '__RJT_CONSTANT__',
                name: 'windowHeight'
              },
              20
            ]
          }
        }
      },
      config
    )
    const { asFragment } = render(element)

    expect(element.props.height).toBe(config.constants.windowHeight as number + 20)
    expect(asFragment()).toMatchSnapshot()
  })
})
