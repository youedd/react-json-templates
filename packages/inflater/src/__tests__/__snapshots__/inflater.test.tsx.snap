// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inflater inflates complex Component 1`] = `
<DocumentFragment>
  <div
    testid="test"
  >
    <button>
       Simple 
    </button>
    <button>
       Simple 
    </button>
  </div>
</DocumentFragment>
`;

exports[`Inflater inflates constants 1`] = `
<DocumentFragment>
  <div
    height="1600"
  >
    900
  </div>
</DocumentFragment>
`;

exports[`Inflater inflates fragment 1`] = `
<DocumentFragment>
  <div>
    <button>
       Simple 
    </button>
    <button>
       Simple 
    </button>
    <button>
       Simple 
    </button>
  </div>
</DocumentFragment>
`;

exports[`Inflater inflates front side calculation 1`] = `
<DocumentFragment>
  <button
    height="1620"
  >
     Simple 
  </button>
</DocumentFragment>
`;

exports[`Inflater inflates props Action 1`] = `
<DocumentFragment>
  <button>
     Simple 
  </button>
</DocumentFragment>
`;

exports[`Inflater inflates simple Component 1`] = `
<DocumentFragment>
  <button
    testid="test"
  >
     Simple 
  </button>
</DocumentFragment>
`;
